File: widget/src/main/java/com/uploadcare/android/widget/adapter/ToolbarSpinnerAdapter.java
package com.uploadcare.android.widget.adapter;

import com.uploadcare.android.widget.R;
import com.uploadcare.android.widget.data.Chunk;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

public class ToolbarSpinnerAdapter extends ArrayAdapter<Chunk> {

    private List<Chunk> mItems = new ArrayList<>();

    private final LayoutInflater inflater;

    public ToolbarSpinnerAdapter(Context context, List<Chunk> objects) {
        super(context, R.layout.ucw_spinner_row, objects);
        mItems = objects;

        inflater = (LayoutInflater) context
                .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    }

    @Override
    public void clear() {
        super.clear();
        mItems.clear();
    }

    public void addItem(Chunk chunk) {
        mItems.add(chunk);
        notifyDataSetChanged();
    }

    public void addItems(List<Chunk> chunks) {
        mItems.addAll(chunks);
        notifyDataSetChanged();
    }

    public void updateItems(List<Chunk> chunks){
        mItems=chunks;
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return mItems.size();
    }

    @Override
    public Chunk getItem(int position) {
        return mItems.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getDropDownView(int position, View view, ViewGroup parent) {
        if (view == null || !view.getTag().toString().equals("DROPDOWN")) {
            view = inflater.inflate(R.layout.ucw_spinner_row, parent, false);
            view.setTag("DROPDOWN");
        }

        TextView textView = (TextView) view.findViewById(R.id.ucw_spinner_title);
        textView.setText(getTitle(position));

        return view;
    }

    @Override
    public View getView(int position, View view, ViewGroup parent) {
        if (view == null || !view.getTag().toString().equals("NON_DROPDOWN")) {
            view = inflater.inflate(R.layout.
                    ucw_toolbar_spinner_item, parent, false);
            view.setTag("NON_DROPDOWN");
        }
        TextView textView = (TextView) view.findViewById(R.id.ucw_spinner_title);
        textView.setText(getTitle(position));
        return view;
    }

    private String getTitle(int position) {
        return position >= 0 && position < mItems.size() ? mItems.get(position).title : "";
    }
}

File: widget/src/main/java/com/uploadcare/android/widget/adapter/ToolbarSpinnerAdapter.kt
package com.uploadcare.android.widget.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.databinding.DataBindingUtil
import com.uploadcare.android.widget.R
import com.uploadcare.android.widget.data.Chunk
import com.uploadcare.android.widget.databinding.UcwSpinnerRowBinding
import com.uploadcare.android.widget.databinding.UcwToolbarSpinnerItemBinding

class ToolbarSpinnerAdapter(context: Context)
    : ArrayAdapter<Chunk>(context, R.layout.ucw_spinner_row) {

    private val chunks = mutableListOf<Chunk>()
    private val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE)
            as LayoutInflater

    override fun clear() {
        super.clear()
        chunks.clear()
    }

    fun updateItems(data: List<Chunk>) {
        chunks.clear()
        chunks.addAll(data)
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return chunks.size
    }

    override fun getItem(position: Int): Chunk? {
        return chunks[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        val binding = convertView?.let {
            DataBindingUtil.getBinding<UcwSpinnerRowBinding>(convertView)
        } ?: UcwSpinnerRowBinding.inflate(inflater, parent, false)

        binding.text = getTitle(position)
        binding.executePendingBindings()

        return binding.root
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val binding = convertView?.let {
            DataBindingUtil.getBinding<UcwToolbarSpinnerItemBinding>(convertView)
        } ?: UcwToolbarSpinnerItemBinding.inflate(inflater, parent, false)

        binding.text = getTitle(position)
        binding.executePendingBindings()

        return binding.root
    }

    private fun getTitle(position: Int): String {
        return chunks[position].title
    }
}